src/invocation/ClusterService.ts

```diff
@@ -286,9 +286,55 @@ export class ClusterService {
     }
 
     private handleMemberList(members: Member[]): void {
+        const prevMembers = this.members;
         this.members = members;
         this.client.getPartitionService().refresh();
         this.logger.info('ClusterService', 'Members received.', this.members);
+        const events = this.detectMembershipEvents(prevMembers);
+        for (const event of events) {
+            this.fireMembershipEvent(event);
+        }
+    }
+
+    private detectMembershipEvents(prevMembers: Member[]): MembershipEvent[] {
+        const events: MembershipEvent[] = [];
+        const eventMembers = Array.from(this.members);
+        const addedMembers: Member[] = [];
+        let deletedMembers = Array.from(prevMembers);
+
+        for (const member of this.members) {
+            const idx = deletedMembers.findIndex(member.equals, member);
+            if (idx === -1) {
+                addedMembers.push(member);
+            } else {
+                deletedMembers = deletedMembers.splice(idx, 1);
+            }
+        }
+
```

Comment:
In the Java client, I see a comment (`// removal events should be added before added events`). I am not sure about the reasoning behind that but maybe we could do the same here

Reply:
Yeah, there must be a good reason behind this particular order. Changed it in 13b65e8db2890eed207d4d88b34de2ff78a2b8e7